packages/react-components/react-table/src/hooks/useSelection.ts

```diff
@@ -1,14 +1,38 @@
-import { GetRowIdInternal, SelectionStateInternal, UseTableOptions } from './types';
-import { useMultipleSelection } from './useMultipleSelection';
-import { useSingleSelection } from './useSingleSelection';
+import * as React from 'react';
+import { useEventCallback, usePrevious } from '@fluentui/react-utilities';
+import { createSelectionManager } from './selectionManager';
+import { GetRowIdInternal, RowId, SelectionMode, SelectionStateInternal } from './types';
 
 export function useSelection<TItem>(
-  selectionMode: UseTableOptions<TItem>['selectionMode'],
+  selectionMode: SelectionMode,
   items: TItem[],
   getRowId: GetRowIdInternal<TItem>,
 ): SelectionStateInternal {
-  const multipleSelectionState = useMultipleSelection(items, getRowId);
-  const singleSelectionState = useSingleSelection();
+  const prevSelectionMode = usePrevious(selectionMode);
+  const [selected, setSelected] = React.useState(() => new Set<RowId>());
+  const [selectionManager, setSelectionManager] = React.useState(() =>
+    createSelectionManager(selectionMode, setSelected),
+  );
 
-  return selectionMode === 'multiselect' ? multipleSelectionState : singleSelectionState;
+  React.useEffect(() => {
+    if (prevSelectionMode !== selectionMode) {
+      setSelectionManager(createSelectionManager(selectionMode, setSelected));
+    }
+  }, [selectionMode, prevSelectionMode]);
+
+  const toggleSelectAllRows: SelectionStateInternal['toggleSelectAllRows'] = useEventCallback(() => {
+    selectionManager.toggleSelectAll(items.map((item, i) => getRowId(item, i)));
+  });
+
+  return {
+    someRowsSelected: selected.size > 0,
+    allRowsSelected: selectionMode === 'single' ? selected.size > 0 : selected.size === items.length,
+    selectedRows: selected,
+    toggleRowSelect: selectionManager.toggleSelect,
+    toggleSelectAllRows,
+    clearSelection: selectionManager.clearSelection,
+    deSelectRow: selectionManager.deSelect,
+    selectRow: selectionManager.select,
+    isRowSelected: selectionManager.isSelected,
+  };
```

Comment:
Wondering about `Row` suffix there. Can cells be selected?

Reply:
Currently that's not a requirement, but it's nice to be explicit and leave `toggleCellSelect` open in the future 🤔